databricks/koalas/groupby.py

```diff
@@ -340,6 +344,103 @@ def any(self):
+
+        .. note:: the output column names are `c0, c1, c2 ... cn`. These names
+            are positionally mapped to the returned DataFrame in ``func``. See examples below.
+
+        .. note:: the dataframe within ``func`` is actually a pandas dataframe. Therefore,
+            any pandas APIs within this function is allowed.
+
+        Parameters
+        ----------
+        func : callable
+            A callable that takes a DataFrame as its first argument, and
+            returns a dataframe.
+
+        Returns
+        -------
+        applied : DataFrame
+
+        See Also
+        --------
+        aggregate : Apply aggregate function to the GroupBy object.
+        Series.apply : Apply a function to a Series.
+
+        Examples
+        --------
+        >>> df = ks.DataFrame({'A': 'a a b'.split(),
+        ...                    'B': [1,2,3],
+        ...                    'C': [4,6, 5]}, columns=['A', 'B', 'C'])
+        >>> g = df.groupby('A')
+
+        Notice that ``g`` has two groups, ``a`` and ``b``.
+        Calling `apply` in various ways, we can get different grouping results:
+
+        Below the functions passed to `apply` takes a DataFrame as
+        its argument and returns a DataFrame. `apply` combines the result for
+        each group together into a new DataFrame:
+
+        >>> def pandas_div_sum(x) -> ks.DataFrame[float, float]:
+        ...    return x[['B', 'C']] / x[['B', 'C']].sum()
+        >>> g.apply(pandas_div_sum)  # doctest: +NORMALIZE_WHITESPACE
+                 c0   c1
+        0  1.000000  1.0
+        1  0.333333  0.4
+        2  0.666667  0.6
+
+        >>> def plus_max(x) -> ks.DataFrame[str, np.int, np.int]:
+        ...    return x + x.max()
+        >>> g.apply(plus_max)  # doctest: +NORMALIZE_WHITESPACE
+           c0  c1  c2
+        0  bb   6  10
+        1  aa   3  10
+        2  aa   4  12
+        """
+        if not isinstance(func, Callable):
+            raise TypeError("%s object is not callable" % type(func))
+
+        assert callable(func), "the first argument should be a callable function."
+        spec = inspect.getfullargspec(func)
+        return_sig = spec.annotations.get("return", None)
+        if return_sig is None:
+            raise ValueError("Given function must have return type hint; however, not found.")
+
+        return_schema = _infer_return_type(func).tpe
+
+        def rename_output(pdf):
+            pdf = func(pdf)
+            # For now, just positionally map the column names to given schema's.
+            pdf = pdf.rename(columns=dict(zip(pdf.columns, return_schema.fieldNames())))
+            return pdf
+
+        grouped_map_func = pandas_udf(return_schema, PandasUDFType.GROUPED_MAP)(rename_output)
+
+        sdf = self._kdf._sdf
+        data_columns = self._kdf._internal.data_columns
+        input_groupkeys = [s._scol for s in self._groupkeys]
+        sdf = sdf.select(data_columns).groupby(*input_groupkeys).apply(grouped_map_func)
```

Comment:
Do we need to `select(data_columns)` here? If needed we should use `data_scols` instead.